docs/advanced/misc.rst

```diff
@@ -39,15 +39,35 @@ The ``PYBIND11_MAKE_OPAQUE`` macro does *not* require the above workarounds.
 Global Interpreter Lock (GIL)
 =============================
 
-When calling a C++ function from Python, the GIL is always held.
+pybind11 will always ensure the Global Interpreter Lock (GIL) is held from
```

Comment:
I think it's great to clarify this section, but I don't think pybind11 does anything actively, which means that "pybind11 will always ensure" isn't accurate.

It's just that there is a general (but not enforced) contract that everyone is meant to adhere to: any thread calling any Python C API function must hold the GIL.

Maybe:

When Python calls into C++ via pybind11, and assuming the process is healthy, the current thread is guaranteed to hold the Global Interpreter Lock (GIL) and pybind11 will never implicitly release it:

```
new code block here
```

The classes :class:`gil_scoped_release` and :class:`gil_scoped_acquire` can be
used to acquire and release the global interpreter lock in the body of a C++
function call. In this way, long-running C++ code can be parallelized using
multiple Python threads, **but great care must be taken** when any :class:`gil_scoped_release` appear: if there is any way that the called C++ code can call back into Python, via callbacks, or trampolines, `gil_scoped_acquire` will be needed.

For callbacks based on `std::function` (pybind11/functional.h), pybind11 will implicitly manage the GIL, most importantly when a Python callback is passed to
C++ code via ``std::function``, and the C++ code calls the function, the GIL is acquired before the Python callback is invoked. Similarly, the `PYBIND11_OVERRIDE` family of macros will acquire the GIL before calling back into Python.